### ISSUE TITLE
[BUG] Memory leak. Memory increases every time trace_method(...) is called

### ISSUE THREAD

**Issue by iNviNho** [DataDog/dd-trace-php#1956]
### Bug description
Hi,

We have been fighting memory leaks in one of our Kafka consumer commands and yesterday we found a cause. 

Every time we consume a new message from Kafka, we start a tracing like this:

```php
        $tracer = new Tracer();
        GlobalTracer::set($tracer);

        $resourceName = $topicName . '::process';

        $options = ['start_time' => Time::now()];
        $rootSpan = $tracer->startRootSpan(
            'Jobs',
            $options
        )
            ->getSpan();

        $rootSpan->setTag(Tag::SPAN_TYPE, Type::CLI);
        $rootSpan->setTag(Tag::SERVICE_NAME, 'broker.service-x');
        $rootSpan->setTag(Tag::RESOURCE_NAME, $resourceName, true);
        $rootSpan->setTag('command_origin', $this->signature);

        // we add tracing for whatever classes we want to trace
        trace_method(Producer::class, 'produce', function (SpanData $span) {
            $span->service = 'broker.offer-service.kafka';
        });
       // ... many more
```

That means that if we consume 10 000 messages in a minute, we will execute this pseudo piece of code 10 000 times. After this block, we added a simple log that prints out current memory usage. **It was increasing every time we executed this block**.

Once we identified that the memory leak is somewhere here, we started removing some lines of code to find out which ones are causing it and we found it. 

**Every time we call *trace_method(...)* MEM usage increases.**

I would expect that the SDK would already make sure that if I call the same trace_method(*) multiple times, **it would ignore all subsequent calls since it sees it as a duplicate**.

### PHP version
8.1

### Tracer version
0.84.0

### OS info
NAME="Alpine Linux"
VERSION_ID=3.17.2
PRETTY_NAME="Alpine Linux v3.17"

**Comment by bwoebi** [comment]
Ouch.

That's not just going to leak memory, but also calling the Closure as many times as it has been registered. ... So it will call the Closure also 10000 times on the 10000th message.

However, it is a feature (and not a bug), that you can call trace_method() multiple times, to evaluate all the Closures. This is particularly important as our own integrations sometimes hook methods which users also hook, and augment with their own data _on top_ of what our integration already sets.

So, the only thing I see an argument for, would be preventing ignoring the same Closure being added to a specific method multiple times.
But the question is, how would you distinguish these? After all `(function() {}) != (function() {})` - each time that code is evaluated, it's a new instance of the Closure.

**Comment by bwoebi** [comment]
To fix your code as-is, you basically have two choices - calling `dd_untrace("produce", Producer::class);` before calling `DDTrace\trace_method()` again or moving the `DDTrace\trace_method` call to a place where it's only called once.

As a mitigation for that issue, and debuggability, I'm wondering whether it would be a good approach to add a limit (100 by default?) to the number of hooks on a particular function and warn when the limit is reached first.

**Comment by bwoebi** [comment]
Hey @iNviNho,

In case you still see such issues, your error log should now contain a log line about repeated hook installing on the same function, with the now released 0.85.0, making you aware of what the problem is.

Let me know if you still experience issues :-)
